### PR TITLE
Improved unit test coverage

### PR DESCRIPTION
**Pull request by azimov**

### CODE REVIEW

File: tests/testthat/setup.R [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Download the JDBC drivers used in the tests
+oldJarFolder <- Sys.getenv("DATABASECONNECTOR_JAR_FOLDER")
+Sys.setenv("DATABASECONNECTOR_JAR_FOLDER" = tempfile("jdbcDrivers"))
+
+withr::defer({
+  unlink(Sys.getenv("DATABASECONNECTOR_JAR_FOLDER"), recursive = TRUE, force = TRUE)
```
Comment (azimov) [inline tests/testthat/setup.R]: @schuemie I took this from Database connector, but i couldn't seem to find a standard HADES environment variable for this in the R_CMD_check configuration in other packages

Comment (schuemie) [reply tests/testthat/setup.R]: With the new version of DatabaseConnector we'll have to start downloading the drivers in all unit tests that need to connect  to these databases. We just haven't done that yet in those other packages.

In DatabaseConnector I used the DATABASECONNECTOR_JAR_FOLDER environmental variable to test that behavior. However, here I would recommend not changing any environmental variables. Instead, we can just specify the `pathToDriver` argument when calling `downloadJdbcDrivers()` and `createConnectionDetails()`. So within the `defer` statement we still need to unlink the temp folder to delete the downloaded drivers. We just need a regular variable holding the temp folder name.

File: tests/testthat/test-Analyses.R [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                                             riskWindowStartUnexposed = -30)
+  analysis2 <- createSccAnalysis(analysisId = 2, description = "test 2", runSelfControlledCohortArgs = args2)
+
+  analysisList <- list(
+    item1 = analysis,
+    item2 = analysis2
+  )
+
+  tmp <- tempfile(fileext = "json")
+  saveSccAnalysisList(analysisList, tmp)
```
Comment (schuemie) [inline tests/testthat/test-Analyses.R:28]: Don't forget to delete (`unlink()`) any temp files once you're done with them.

File: tests/testthat/test-RunAnalyses.R [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  runSelfControlledCohortArgs2 <- createRunSelfControlledCohortArgs(firstExposureOnly = TRUE)
+  sccAnalysis1 <- createSccAnalysis(analysisId = 1,
+                                    runSelfControlledCohortArgs = runSelfControlledCohortArgs1)
+  sccAnalysis2 <- createSccAnalysis(analysisId = 2,
+                                    runSelfControlledCohortArgs = runSelfControlledCohortArgs2)
+  sccAnalysisList <- list(sccAnalysis1, sccAnalysis2)
+
+  outputFolder <- file.path(tempdir(), getOption("dbms"))
+  dir.create(outputFolder)
+  on.exit(unlink(outputFolder))
```
Comment (schuemie) [inline tests/testthat/test-RunAnalyses.R]: Could you confirm that the `on.exit` clause gets executed at the end of the unit test? (I've only used it inside functions, so I'm not sure it works in this context)

File: .github/workflows/R_CMD_check_Hades.yaml [hunk 4]
```diff
@@ -77,20 +107,86 @@ jobs:
...
+          name: package_tarball    
+          
+      - name: Push to drat
+        if: ${{ env.new_version != '' }}
+        run: |
+          bash deploy.sh
+      - name: Push to BroadSea
+        if: ${{ env.new_version != '' }}
         run: |
-          perl compare_version --tag && \
```
Comment (ablack3) [inline .github/workflows/R_CMD_check_Hades.yaml:94]: Hi @azimov,  Is compare_versions used to create a github release? Are we removing this from all HADES packages?

File: tests/testthat/test-Analyses.R [hunk 5]
```diff
@@ -0,0 +1,71 @@
+library(testthat)
+
+test_that("createSccAnalysis, saveSccAnalysisList, loadSccAnalysisList", {
+  args1 <- createRunSelfControlledCohortArgs(riskWindowStartUnexposed = -60)
+
+  analysis <- createSccAnalysis(analysisId = 1, description = "test", runSelfControlledCohortArgs = args1)
+
+  expect_equal(class(analysis), "sccAnalysis")
```
Comment (ablack3) [inline tests/testthat/test-Analyses.R]: Another option is `expect_s3_class(analysis, "sccAnalysis")`

File: tests/testthat/test-Analyses.R [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+  expect_error(saveSccAnalysisList(c(1,2), tmp)) # Not a list
+  expect_error(saveSccAnalysisList(list(), tmp)) # Empty
+  expect_error(saveSccAnalysisList(badAnalysisList, tmp)) # Not an sccAnalysis object
+})
+
+
+test_that("loadExposureOutcomeList, saveExposureOutcomeList , createExposureOutcome", {
+  exposureOutcome1 <- createExposureOutcome(123, 456)
+  expect_is(exposureOutcome1, "exposureOutcome")
```
Comment (ablack3) [inline tests/testthat/test-Analyses.R]: Apparently `expect_is` is superceded and deprecated in [the 3rd edition](https://testthat.r-lib.org/articles/third-edition.html) and [the documentation ](https://rdrr.io/cran/testthat/man/expect_is.html)recommends using `expect_s3_class(exposureOutcome1, "exposureOutcome")` instead. I think `expect_is` will continue to work in the older edition of test_that though.

File: tests/testthat/test-RunAnalyses.R [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+                       outputFolder = outputFolder,
+                       computeThreads = 8)
+
+  expect_is(rr, "data.frame")
+  expect_true(file.exists(file.path(outputFolder, "resultsReference.rds")))
+  apply(rr, 1, function(item) {
+    expect_true(file.exists(file.path(outputFolder, item["sccResultsFile"])))
+  })
+
+  result <- summarizeAnalyses(rr, outputFolder)
```
Comment (ablack3) [inline tests/testthat/test-RunAnalyses.R:38]: Should you also check that the actual results in some way? Something like `expect_true(all(result$irr > 0.1 & result$irr < 5))`

**Review by ablack3 — APPROVED** [review]
•  Do these tests meet the code coverage requirement? Yes
•  Does the PR pass all github actions checks without failure? Yes
•  For the SQL that is called: is it actually being executed or have we just called it with a bunch of null data? The examples are realistic.
•  Is the code tested on multiple database platforms? Yes 
•  Is the data that is being used to test the SQL the kind of data that we would expect to find in a real CDM? I think it is synpuf so sort of. 
•  Do the tests make sense or do they just get coverage of lines of code. Yes they make sense.
•  Do the tests cover any edge cases that you can think of? One edge case that might come up fairly often is when the number of persons in the cohort is 0. In this case we get an sccResults object with an estimates dataframe with 0 rows. When we summarize several analyses the comparisons with 0 persons are excluded and if all analyses have 0 persons then we get a dataframe with 0 rows and 0 columns. This does not seem ideal to me because it is kind of like type instability. Even though the output of  summarizeAnalyses() is always a dataframe the number of columns of that dataframe can be be 0 or 14. A couple things I think we could improve in HADES are using type consistent functions and separating side effects (actions) from calculations (pure functions). 
•  Is the code in the tests up to the HADES developer guidelines? It looks like you are removing compare_version which is included in the guidelines. Otherwise yes.
•  Are there better/more efficient ways to do these tests? Not that I can see.
•  Would these tests help someone new to the package understand the package? With some explanation maybe but I think this question is more relevant for vignettes. Tests could be added to a vignette as examples though.
•  Can this checklist be improved? 
	“Is the data that is being used to test the SQL the kind of data that we would expect to find in a real CDM?” The data will always be synpuf I think. A more relevant question would be “Is the test code realistic?”
“Would these tests help someone new to the package understand the package?” – I’d consider dropping this question because the purpose of the tests is to prevent breaking code. However the code should be readable and understandable so maybe it could be revised to “Is the code written clearly and in a way that the intention of the test is understandable?” Something like that.

Awesome work. It was helpful for me to read through your tests.

**Comment by msuchard** [comment]
@azimov -- thanks for the hard work.
@ablack3 -- great work with the review.

File: .github/workflows/R_CMD_check_Hades.yaml (hunk 4 above)
Comment (azimov) [reply .github/workflows/R_CMD_check_Hades.yaml:94]: Good spot, not sure why this got removed, probably from a copy paste!

File: tests/testthat/test-RunAnalyses.R (hunk 7 above)
Comment (azimov) [reply tests/testthat/test-RunAnalyses.R:38]: Many results will be Inf as there will by outputs with 0 exposures and 0 outcomes. I will try and think of a way to test the actual process with hand generated in a future version.

**Comment by azimov** [comment]
Thanks for the review @ablack3 

>  • Do the tests cover any edge cases that you can think of? One edge case that might come up fairly often is when the number of persons in the cohort is 0. In this case we get an sccResults object with an estimates dataframe with 0 rows. When we summarize several analyses the comparisons with 0 persons are excluded and if all analyses have 0 persons then we get a dataframe with 0 rows and 0 columns. This does not seem ideal to me because it is kind of like type instability. Even though the output of summarizeAnalyses() is always a dataframe the number of columns of that dataframe can be be 0 or 14. A couple things I think we could improve in HADES are using type consistent functions and separating side effects (actions) from calculations (pure functions).

Thanks, this is a really important point. The outputted data frame should have consistent columns even if it has 0 data because cohorts are empty. I have made changes that ensure that the columns are consistent. This also helped me stop a potential bug where multiple analyses could have different resulting columns if they had the `computeTarDistribution` flag set differently. The outputs now have guaranteed consistent columns.

**Comment by ablack3** [comment]
> Thanks, this is a really important point. The outputted data frame should have consistent columns even if it has 0 data because cohorts are empty. I have made changes that ensure that the columns are consistent. This also helped me stop a potential bug where multiple analyses could have different resulting columns if they had the `computeTarDistribution` flag set differently. The outputs now have guaranteed consistent columns.

Cool! Glad to know I’m not being overly nitpicky.
